src/ResumableStream.php

```diff
@@ -96,34 +117,41 @@ public function __construct(
      */
     public function readAll()
     {
-        $tries = 0;
-        $argumentFunction = $this->argumentFunction;
-        $retryFunction = $this->retryFunction;
+        // Reset $currentAttempts on successful row read, but keep total attempts for the header.
+        $currentAttempt = $totalAttempt = 0;
         do {
             $ex = null;
-            $args = $argumentFunction();
-            if (!isset($args[1]['requestCompleted']) || $args[1]['requestCompleted'] !== true) {
-                if ($tries > 0) {
-                    // Send in "bigtable-attempt" header on retry
-                    $optionalArgs = array_pop($args);
+            $args = ($this->argumentFunction)();
+
+            $completed = ($args[1]['requestCompleted'] ?? false) === true;
+            if ($completed !== true) {
+                $optionalArgs = array_pop($args);
+                $optionalArgs['retrySettings'] = ['retriesEnabled' => false]; // disable retries in GAPIC layer
+                // Send in "bigtable-attempt" header on retry request
+                if ($totalAttempt > 0) {
                     $headers = $optionalArgs['headers'] ?? [];
-                    $headers['bigtable-attempt'] = [(string) $tries];
+                    $headers['bigtable-attempt'] = [(string) $totalAttempt];
                     $optionalArgs['headers'] = $headers;
-                    $args[] = $optionalArgs;
+                    ($this->delayFunction)($currentAttempt);
                 }
-                $stream = $this->createExponentialBackoff()->execute($this->apiFunction, $args);
+                $args[] = $optionalArgs;
+
+                $stream = call_user_func_array($this->apiFunction, $args);
+
                 try {
                     foreach ($stream->readAll() as $item) {
                         yield $item;
+                        $currentAttempt = 0; // reset delay and attempt on successful read.
                     }
                 } catch (\Exception $ex) {
                     if ($this->logger) {
                         $this->logger->error($ex->getMessage());
                     }
                 }
+                $totalAttempt++;
+                $currentAttempt++;
             }
-            $tries++;
-        } while ((!$this->retryFunction || $retryFunction($ex)) && $tries <= $this->retries);
+        } while ((!$this->retryFunction || ($this->retryFunction)($ex)) && $currentAttempt <= $this->retries);
```

Comment:
I think this is correct but really hard to read because of variable naming and shadowing. In a follow up pr, I think it would make sense to update this to:

```php
$callerArgs = ($this->argumentFunction)();
$tableName = $callerArgs[0];
$requestOpts = $callerArgs[1];

$skipRequest = ($requestOpts["requestCompleted"] ... ;
if ($skipRequest) { ...}
//..
$requestOpts['headers'] = ....
// ...
$stream = call_user_func(this->apiFunction, $tableName, $requestOpts)
```

Reply:
I do agree the v1 code here is incredibly hard to decipher. One of the issues with V1 is that we can't be sure which argument to `$callerArgs` is the `$requestOpts` because depending on the RPC, the number of arguments change. For example, if the RPC being called is something other than `readRows`, it's possible that `$callerArgs[0]` will not be _not_ `$tableName`. At least in the case of `$requestOpts`, it's better to use `array_pop`, as the last argument should always be the `$optionalArgs` array.

Although, since that's already hardcoded, I guess this issue has already been introduced.